From f087b9fd10ae13485c0b2bdc1a613503b3da360c Mon Sep 17 00:00:00 2001
From: Runfeng Xiao <runfeng.xiao@amlogic.com>
Date: Sat, 4 Jun 2022 16:45:21 +0800
Subject: [PATCH 21/30] gstreamer: reduce queue size and refine read blocksize
 [1/2]

PD#TV-53466

Problem:
Crash due to use huge memory.

Solution:
reduce queue size and refine read blocksize

Verify:
on t5d am32a2

Change-Id: I7d4b762d5d30285f979c20425010d9d5ffd83046
---
 ext/hls/gsthlsdemux-util.c                    | 5 +++++
 ext/hls/gsthlsdemux.c                         | 2 --
 gst-libs/gst/adaptivedemux/gstadaptivedemux.c | 6 ++++--
 3 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/ext/hls/gsthlsdemux-util.c b/ext/hls/gsthlsdemux-util.c
index 95f61d4..8bac2f6 100644
--- a/ext/hls/gsthlsdemux-util.c
+++ b/ext/hls/gsthlsdemux-util.c
@@ -208,6 +208,11 @@ gst_hlsdemux_tsreader_find_pcrs_mpegts (GstHLSTSReader * r,
   const guint8 *data;
   gsize size;
 
+  *first_pcr = r->first_pcr = 0;
+  *last_pcr = r->last_pcr = 0;
+
+   return TRUE;
+
   if (!gst_buffer_map (buffer, &info, GST_MAP_READ))
     return FALSE;
 
diff --git a/ext/hls/gsthlsdemux.c b/ext/hls/gsthlsdemux.c
index 4b6109f..cee3f76 100755
--- a/ext/hls/gsthlsdemux.c
+++ b/ext/hls/gsthlsdemux.c
@@ -869,7 +869,6 @@ gst_hls_demux_handle_buffer (GstAdaptiveDemux * demux,
     hls_stream->pending_pcr_buffer = NULL;
   }
 
-#if 0 //delete for airplay youtube play slowly
   if (!gst_hlsdemux_tsreader_find_pcrs (&hls_stream->tsreader, &buffer,
           &first_pcr, &last_pcr, &tags)
       && !at_eos) {
@@ -877,7 +876,6 @@ gst_hls_demux_handle_buffer (GstAdaptiveDemux * demux,
     hls_stream->pending_pcr_buffer = buffer;
     return GST_FLOW_OK;
   }
-#endif
 
   if (tags) {
     gst_adaptive_demux_stream_set_tags (stream, tags);
diff --git a/gst-libs/gst/adaptivedemux/gstadaptivedemux.c b/gst-libs/gst/adaptivedemux/gstadaptivedemux.c
index 543e108..11d996a 100644
--- a/gst-libs/gst/adaptivedemux/gstadaptivedemux.c
+++ b/gst-libs/gst/adaptivedemux/gstadaptivedemux.c
@@ -130,7 +130,9 @@ GST_DEBUG_CATEGORY (adaptivedemux_debug);
 #define DEFAULT_FAILED_COUNT 3
 #define DEFAULT_CONNECTION_SPEED 0
 #define DEFAULT_BITRATE_LIMIT 0.8f
-#define SRC_QUEUE_MAX_BYTES 20 * 1024 * 1024    /* For safety. Large enough to hold a segment. */
+#define SRC_QUEUE_MAX_BYTES 8 * 1024 * 1024    /* For safety. Large enough to hold a segment. */
+#define SRC_QUEUE_MAX_BUFS 3000
+
 #define NUM_LOOKBACK_FRAGMENTS 3
 
 #define GST_MANIFEST_GET_LOCK(d) (&(GST_ADAPTIVE_DEMUX_CAST(d)->priv->manifest_lock))
@@ -3044,7 +3046,7 @@ gst_adaptive_demux_stream_update_source (GstAdaptiveDemuxStream * stream,
       return FALSE;
 
     g_object_set (queue, "max-size-bytes", (guint) SRC_QUEUE_MAX_BYTES, NULL);
-    g_object_set (queue, "max-size-buffers", (guint) 0, NULL);
+    g_object_set (queue, "max-size-buffers", (guint) SRC_QUEUE_MAX_BUFS, NULL);
     g_object_set (queue, "max-size-time", (guint64) 0, NULL);
 
     uri_handler = gst_element_make_from_uri (GST_URI_SRC, uri, NULL, NULL);
-- 
2.34.1

